### ISSUE TITLE
scipy.signal.stft (and istft)

### ISSUE THREAD

**Issue by yotarok** [google/jax#9411]
STFT plays a vital role in audio/ speech signal processing, and it would be nice to have jax-version of `scipy.signal.stft` function.
As Jax already has auto-differentiable FFT and conv1d, it can be basically a utility function that combines those two.

It would be nice if we can have [SciPy-compatible](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html#scipy.signal.stft) STFT function in Jax.

**Comment by machineko** [comment]
Hey you can use some old audio prepro i created (librosa compatible) https://github.com/machineko/jaxaudio/blob/7320a4ecd8f267e206c36be3551f919faa08d3b0/jax_audio/utils/audio_utils.py#L128 
To be compatible with scipy you could just need to implemented scaling (i think its only difference outside method arguments that can be renamed) Feel free to use it or if u have time to reimplement it into core jax repo

**Comment by yotarok** [comment]
Hi machineko,

Thanks for the pointer.
Yes, as you showed in the example, implementation of STFT shouldn't be that complicated.
So, my the request here is more about asking whether Jax team is comfortable with having STFT implementation in the repo.
It's nice to have a standard place where we can merge our efforts, especially because STFT is a part of scipy.signal.

**Comment by jakevdp** [comment]
I don't know of any work on this, but I think the JAX team would be happy to accept a contribution of a wrapper for `scipy.signal.stft` within `jax.scipy.signal`

**Comment by machineko** [comment]
@jakevdp Are jax functions should mimic scipy fully? 
Or can we make some more modern approach for that (and dont follow full scipy convention for example change convention of outputs/inputs?)

**Comment by hawkinsp** [comment]
If JAX functions are named like `jax.scipy.XYZ` they should try to mimic the `scipy` API. If they diverge, then that's fine, but they should be called something else.

**Comment by machineko** [comment]
I was looking to implement api compatible with both librosa and scipy (with some mode parameter and limited functionality) but with more "librosa like" input and output convention (not creating freq and time https://github.com/scipy/scipy/blob/47bb6febaa10658c72962b9615d5d5aa2513fa3a/scipy/signal/spectral.py#L1852 ) 
also using auto-calculate padding with base reflect mode (in scipy this is manual with boundary mode)

But i can agree it shouldn't be call scipy.sftf as it wouldn't mimic all of the scipy functions, so maybe just jax.sftf?

**Comment by yotarok** [comment]
Hi @machineko, and @hawkinsp,

I think having the extra outputs in "scipy.stft" is not significan because JIT can effectively omit computation of those if we just didn't use those outputs.
And, "reflect"-paddings in librosa are implemented in scipy.stft too when boundary="even" is provided.

So, overall, I prefer strictly maintaining scipy.stft interface if we are to put the code in Jax repo.
Here's a proposal PR.
https://github.com/google/jax/pull/9422

Any feedbacks are welcome!

**Comment by psmaragdis** [comment]
Any chance we also get scipy.signal.istft?  Or I'd be happy to receive pointers on how one can make a jitable implementation. The output array is dependent on the function parameters so the usual approaches result in a TracerIntegerConversionError. For the stft you get around this by using jax.lax.conv_general_dilated_patches, but I'm not sure what the equivalent function would be for an istft implementation.

**Comment by yotarok** [comment]
I looked into this.

It might be less computationally efficient, but overlap-sum can be done with dilated convolution and a kernel that represents time-shift.
I wrote a (not-working-yet) initial version using this strategy: https://github.com/yotarok/jax/blob/istft/jax/_src/scipy/signal.py#L580

Hopefully, I can fix the current glitches, and create a PR containing this, soon.

**Comment by yotarok** [comment]
Sent the PR.

I found that the method I proposed above (with dilated convolution) was not scalable to realistic sizes of input signals.
Instead, I used a for-loop that will be iterated depending on the overlap ratio.
While this also has an inefficiency when overlap ratio is very high, I believe this is okay for common use cases.

**Comment by hawkinsp** [comment]
I haven't thought about this in any detail, but the TF implementation has another implementation that may be worth looking at also:
https://github.com/tensorflow/tensorflow/blob/87462bfac761435a46641ff2f10ad0b6e5414a4b/tensorflow/python/ops/signal/spectral_ops.py#L157
and
https://github.com/tensorflow/tensorflow/blob/87462bfac761435a46641ff2f10ad0b6e5414a4b/tensorflow/python/ops/signal/reconstruction_ops.py#L28

**Comment by yotarok** [comment]
Thanks Peter, TF's implementation looks better than the for-loop-based one.
I implemented the utility function following this way, and updated the PR.
